### ISSUE TITLE
tf.nn.softmax loses a finite first-derivative signal for float64 input

### ISSUE THREAD

**Issue by ALinrunrun** [tensorflow/tensorflow#126633]
### Issue type

Bug

### Have you reproduced the bug with TensorFlow Nightly?

Yes

### Source

source

### TensorFlow version

tf 2.21.0

### Custom code

Yes

### OS platform and distribution

Linux Ubuntu 22.04

### Python version

Python 3.13

### Current behavior?

`tf.nn.softmax` returns zero for one first-derivative component whose correctly rounded `float64` reference value is finite, normal, and nonzero.

Softmax is invariant to adding the same constant to every logit, so every input VJP should sum to zero. In the reproducer below, the expected gradient is approximately `[5.551115123125775e-17, -5.551115123125775e-17]`.

However, TensorFlow returns `[0.0, -5.551115123125776e-17]`. The first gradient component is incorrectly collapsed to zero, and the resulting VJP no longer preserves the expected zero-sum relation.

Expected behavior?

The first-order autodiff result for `tf.nn.softmax` should preserve the finite, normal, nonzero `float64` gradient component. For the given input and cotangent, the gradient should be close to `[5.551115123125775e-17, -5.551115123125775e-17]`, and the VJP should sum to zero.

### Standalone code to reproduce the issue

```shell
import os
os.environ.setdefault("CUDA_VISIBLE_DEVICES", "-1")
os.environ.setdefault("TF_CPP_MIN_LOG_LEVEL", "3")

import tensorflow as tf

dtype = tf.float64

expected = [5.551115123125775e-17, -5.551115123125775e-17]

z = tf.Variable([37.42994775023705, 0.0], dtype=dtype)
cotangent = tf.constant([1.0, 0.0], dtype=dtype)

with tf.GradientTape() as tape:
    y = tf.nn.softmax(z, axis=-1)
    objective = tf.reduce_sum(y * cotangent)

actual = tape.gradient(objective, z)
vjp_sum = tf.reduce_sum(actual)

print("actual:")
print(actual.numpy().tolist())
print("expected:")
print(expected)
print("VJP sum:")
print(float(vjp_sum))

if actual.numpy()[0] == 0.0 and expected[0] != 0.0:
    print("BUG REPRODUCED: tf.nn.softmax loses a finite first-derivative signal")
else:
    print("not reproduced")
```

### Relevant log output

```shell
actual:
[0.0, -5.551115123125776e-17]
expected:
[5.551115123125775e-17, -5.551115123125775e-17]
VJP sum:
-5.551115123125776e-17
BUG REPRODUCED: tf.nn.softmax loses a finite first-derivative signal
```

**Comment by Kayyuri** [comment]
Hi @ALinrunrun, Thank you for reporting this detailed issue! You've clearly demonstrated that `tf.nn.softmax`'s gradient calculation for `float64` inputs appears to incorrectly collapse a finite, non-zero derivative component to zero, which then causes the VJP sum to deviate from the expected zero. The provided standalone code effectively reproduces the bug. We appreciate you bringing this to our attention.

**Comment by Kayyuri** [comment]
Hi @ALinrunrun,
A fix for this issue has been proposed in this [PR](https://github.com/tensorflow/tensorflow/pull/126823). Once it is merged, the issue will be resolved.

Many thanks to @GodlyDonuts for raising the PR to address this issue.
